interpreter/beam/beam.go

```diff
@@ -0,0 +1,164 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package beam // import "go.opentelemetry.io/ebpf-profiler/interpreter/beam"
+
+// BEAM VM Unwinder support code
+
+// The BEAM VM is an interpreter for Erlang, as well as several other languages
+// that share the same bytecode, such as Elixir and Gleam.
+
+import (
+	"fmt"
+	"regexp"
+	"unsafe"
+
+	"go.opentelemetry.io/ebpf-profiler/internal/log"
+
+	"go.opentelemetry.io/ebpf-profiler/host"
+	"go.opentelemetry.io/ebpf-profiler/interpreter"
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/lpm"
+	"go.opentelemetry.io/ebpf-profiler/process"
+	"go.opentelemetry.io/ebpf-profiler/remotememory"
+	"go.opentelemetry.io/ebpf-profiler/reporter"
+	"go.opentelemetry.io/ebpf-profiler/support"
+)
+
+var (
+	// regex for matching the process name
+	beamRegex                      = regexp.MustCompile(`(^|\/)beam\.smp`)
+	_         interpreter.Data     = &beamData{}
+	_         interpreter.Instance = &beamInstance{}
+)
+
+type beamData struct {
+	otpRelease  string
+	ertsVersion string
+}
+
+type beamInstance struct {
+	interpreter.InstanceStubs
+
+	pid  libpf.PID
+	data *beamData
+	rm   remotememory.RemoteMemory
+	bias libpf.Address
+
+	// prefixes is indexed by the prefix added to ebpf maps (to be cleaned up) to its generation
+	prefixes map[lpm.Prefix]uint32
+	// mappingGeneration is the current generation (so old entries can be pruned)
+	mappingGeneration uint32
+}
+
+func Loader(ebpf interpreter.EbpfHandler, info *interpreter.LoaderInfo) (interpreter.Data, error) {
+	matches := beamRegex.FindStringSubmatch(info.FileName())
+	if matches == nil {
+		return nil, nil
+	}
+
+	ef, err := info.GetELF()
+	if err != nil {
+		return nil, err
+	}
+
+	_, otpRelease, err := ef.SymbolData("etp_otp_release", 4)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read OTP release: %v", err)
+	}
+
+	_, ertsVersion, err := ef.SymbolData("etp_erts_version", 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read ERTS version: %v", err)
+	}
+
+	d := &beamData{
+		otpRelease:  string(otpRelease),
+		ertsVersion: string(ertsVersion),
+	}
+
+	return d, nil
+}
+
+func (d *beamData) Attach(ebpf interpreter.EbpfHandler, pid libpf.PID, bias libpf.Address, rm remotememory.RemoteMemory) (interpreter.Instance, error) {
```

Comment:
Can you also add a `String()` decorator for the `beamData`? It is used in few places to log things. See also 278b85e5c71f142fa8d86928e116762a78a84227